### PR TITLE
[Program Capture] Add pytree support to captured `qml.grad` and `qml.jacobian`

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
#6120  and #6127 add support to capture `qml.grad` and `qml.jacobian` in plxpr. Once captured, they dispatch to `jax.grad` and `jax.jacobian`.

**Description of the Change:**
This PR adds support for pytree inputs and outputs of the differentiated functions, similar to #6081.
For this, it extends the internal class `FlatFn` by the extra functionality to turn the wrapper into a `*flat_args -> *flat_outputs` function, instead of a `*pytree_args -> *flat_outputs` function.

**Benefits:**
Pytree support :deciduous_tree: 

**Possible Drawbacks:**

**Related GitHub Issues:**

[sc-70930]
[sc-71862]

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -26,8 +26,10 @@
   into plxpr. When evaluating a captured `qml.grad` (`qml.jacobian`) instruction, it will
   dispatch to `jax.grad` (`jax.jacobian`), which differs from the Autograd implementation
   without capture.
+  Pytree inputs and outputs are supported.
```
Comment (mudit2812) [inline doc/releases/changelog-dev.md:29]:
```suggestion
  without capture. Pytree inputs and outputs are supported.
```

File: pennylane/_grad.py [hunk 2]
```diff
@@ -42,9 +45,40 @@ def _capture_diff(func, argnum=None, diff_prim=None, method=None, h=None):
...
+            (
+                (
+                    list(range(start, (start := start + len(flat_arg))))
+                    if i in argnum
+                    else list(range((start := start + len(flat_arg)), start))
+                )
+                for i, flat_arg in enumerate(flat_args)
+            ),
+            start=[],
+        )
```
Comment (mudit2812) [inline pennylane/_grad.py:68]: ~What's the point of setting `start=[]`?~ Did some local coding and I see why it's used. Very cool, and concise solution, though I found it a bit unreadable. I'd say `black` is partially to blame. Maybe it'll be a bit more readable if you split it up into one step to create the generator, and another to compute its sum?
```suggestion
        start = 0
        flat_argnum_gen = (
            (
                list(range(start, (start := start + len(flat_arg))))
                if i in argnum
                else list(range((start := start + len(flat_arg)), start))
            )
            for i, flat_arg in enumerate(flat_args)
        )
        flat_argnum = sum(flat_argnum_gen, start=[])
```

File: pennylane/capture/flatfn.py
Comment (mudit2812) [inline pennylane/capture/flatfn.py:1]: Thoughts on adding `FlatFn` to `qml.capture.__init__.py`? The docstring looks great, but it's not viewable from the docs. The `capture` module is already under the "Internals" section, so I don't personally see an issue with exposing this class there.

Not needed in this PR though.

File: tests/capture/test_capture_diff.py [hunk 3]
```diff
@@ -473,3 +521,63 @@ def circuit(x):
...
+        def inner_func(x, y):
+            return {
+                "prod_cos": jnp.prod(jnp.sin(x["a"]) * jnp.cos(y[0]["b"][1]) ** 2),
+                "sum_sin": jnp.sum(jnp.sin(x["a"]) * jnp.sin(y[1]["c"]) ** 2),
+            }
+
+        def func_qml(x):
+            return qml.jacobian(inner_func, argnum=argnum)(
+                {"a": x}, ({"b": [None, 0.4 * jnp.sqrt(x)]}, {"c": 0.5})
+            )
```
Comment (mudit2812) [inline tests/capture/test_capture_diff.py:543]: I love that this is working 😄

**Review by mudit2812 — APPROVED** [review]
Thanks @dwierichs . Looks great 🎉 🚀

File: pennylane/_grad.py (hunk 2 above)
Comment (dwierichs) [reply pennylane/_grad.py:68]: Good idea! :) Yeah I must admit it is not the most readable expression :thinking:

File: pennylane/capture/flatfn.py
Comment (dwierichs) [reply pennylane/capture/flatfn.py:1]: I like the idea! Will wait for @albi3ro to comment :)

Comment (albi3ro) [reply pennylane/capture/flatfn.py:1]: I'm fine either way.

File: pennylane/capture/capture_diff.py [hunk 4]
```diff
@@ -103,7 +103,7 @@ def _(*args, argnum, jaxpr, n_consts, method, h):
         def func(*inner_args):
             return jax.core.eval_jaxpr(jaxpr, consts, *inner_args)
 
-        return jax.jacobian(func, argnums=argnum)(*args)
+        return jax.tree_util.tree_flatten(jax.jacobian(func, argnums=argnum)(*args))[0]
```
Comment (albi3ro) [inline pennylane/capture/capture_diff.py:106]:
```suggestion
        return jax.tree_util.tree_leaves(jax.jacobian(func, argnums=argnum)(*args))
```
Maybe?

**Review by albi3ro — COMMENTED** [review]
Am I missing something? 

```
@qml.qnode(qml.device('default.qubit', wires=2))
def circuit(x, y):
    qml.RX(x,0)
    qml.RY(y['a'], 0)
    qml.RX(y['b'], 0)
    return qml.expval(qml.Z(0))

qml.grad(circuit)(jax.numpy.array(0.5), {"a":jax.numpy.array(1.0), "b": jax.numpy.array(2.0)})
```
```
TypeError: primal and tangent arguments to jax.jvp must have the same tree structure; primals have tree structure PyTreeDef((*, *, *)) whereas tangents have tree structure PyTreeDef((*, CustomNode(Zero[ShapedArray(float32[], weak_type=True)], []), CustomNode(Zero[ShapedArray(float32[], weak_type=True)], []))).
```

**Comment by dwierichs** [comment]
The issue seems to be that defining a primitive's JVP rule using `jax.jvp` again exposes the zero tracers sent through the program as tangents to the custom implementation, which breaks when looking at the `PytreeDef`s, because zeros are registered as `CustomNode`. [This issue existed with a previous implementation](https://github.com/google/jax/issues/1670) in JAX, back in 2019. The old issue re-appears with up-to-date JAX code :thinking: Should I post an issue on the JAX repo for this?

File: pennylane/capture/flatfn.py
Comment (dwierichs) [reply pennylane/capture/flatfn.py:1]: Added it to the module.

**Review by mudit2812 — APPROVED** [review]
Approving again 😄

**Review by albi3ro — APPROVED** [review]
👍

**Pull request merged**